src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -46,34 +50,75 @@ public static void generateReposReport(List<RepoConfiguration> configs, String o
+            RepoLocation location = config.getLocation();
+            if (repoLocationMap.containsKey(location)) {
+                repoLocationMap.get(location).add(config);
+            } else {
+                List<RepoConfiguration> configsList = new ArrayList<>();
+                configsList.add(config);
+                repoLocationMap.put(location, configsList);
+            }
+        }
+        return repoLocationMap;
+    }
+
+    /**
+     * Clone, analyze and generate the report for repositories in {@code repoLocationMap}.
      * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
      */
-    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) throws IOException {
+    private static void cloneAndAnalyzeRepos(
+            Map<RepoLocation, List<RepoConfiguration>> repoLocationMap, String outputPath) throws IOException {
         RepoCloner repoCloner = new RepoCloner();
-        RepoConfiguration clonedRepo = null;
+        RepoLocation clonedRepoLocation = null;
 
-        for (RepoConfiguration config : configs) {
-            repoCloner.clone(outputPath, config);
+        for (RepoLocation location : repoLocationMap.keySet()) {
+            repoCloner.clone(outputPath, repoLocationMap.get(location).get(0));
 
-            if (clonedRepo != null) {
-                analyzeRepo(outputPath, clonedRepo);
+            if (clonedRepoLocation != null) {
+                analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation), repoCloner);
             }
-            clonedRepo = repoCloner.getClonedRepo(outputPath);
+            clonedRepoLocation = repoCloner.getClonedRepo(outputPath);
         }
-        if (clonedRepo != null) {
-            analyzeRepo(outputPath, clonedRepo);
+        if (clonedRepoLocation != null) {
+            analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation), repoCloner);
         }
         repoCloner.cleanup();
     }
 
+    /**
+     * Analyzes repos in {@code configs} and generates the report.
+     */
+    private static void analyzeRepos(
+            String outputPath, List<RepoConfiguration> configs, RepoCloner repoCloner) throws IOException {
+
+        for (RepoConfiguration config : configs) {
+            try {
+                repoCloner.updateAndCheckoutBranch(config);
+            } catch (RuntimeException e) {
+                logger.log(Level.SEVERE, "Branch does not exist! Analysis terminated.", e);
+                Path repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
+                FileUtil.createDirectory(repoReportDirectory);
+                generateEmptyRepoReport(repoReportDirectory.toString());
+                return;
+            }
+            analyzeRepo(outputPath, config);
+        }
+    }
+
     /**
      * Analyzes repo specified by {@code config} and generates the report.
      */
```

Comment:
I just noticed this, seems like `analyzeRepo` isn't catching the `RunTimeException` at the correct place, should be with the bottom chunk right? And also have to handle creating empty report for there?

Reply:
Yeah this seems weird. I went to take a look at the history before cloning in parallel [here](https://github.com/reposense/RepoSense/blob/d426f09a3a5271d306c751803ec0d8ccabfdb87c/src/main/java/reposense/report/ReportGenerator.java).

For the bottom chunk, it seems like there wasn't any catch block for `RunTimeException` there all along. For the top chunk, the `RuntimeException` catch block was introduced in this [commit](https://github.com/reposense/RepoSense/commit/3d393c513a262d612629bf2eb6511cd04d6bad87) and so I just kept it there.

From the history, it also seems like in the case of `IOException` and `RuntimeException` we just skip the repo entirely and didn't create an empty report.

Reply:
I can change it if you think its a mistake though.

Reply:
>I can change it if you think its a mistake though.

This seems outside of the scope of this PR, so let's leave it for now. We can raise an issue and resolve this quickly if problems occur in CS2103/CS2113.